src/cargo/ops/common_for_install_and_uninstall.rs

```diff
@@ -520,17 +519,23 @@ pub fn path_source(source_id: SourceId, config: &Config) -> CargoResult<PathSour
     Ok(PathSource::new(&path, source_id, config))
 }
 
+/// Either the specific dependency to be selected, or a function which can be executed to list all
+/// packages known in a source so that one can be selected from that list.
+pub enum DependencyOrListAllFn<F> {
+    Dependency(Dependency),
+    ListAllFn(F),
+}
+
 /// Gets a Package based on command-line requirements.
-pub fn select_pkg<'a, T>(
-    mut source: T,
-    name: Option<&str>,
-    vers: Option<&str>,
+pub fn select_pkg<T, F>(
+    source: &mut T,
+    dep_or_list_all_fn: DependencyOrListAllFn<F>,
```

Comment:
I'm a bit confused why this enum needs to exist.  Can there just be two different `select_pkg` functions?  I don't see any calls to `select_pkg` where it doesn't know which variant to use.